### PR TITLE
Fetching Nonce from getBalance

### LINKED ISSUE #116: Add Nonce-functionality for Transaction
In order to support https://github.com/0chain/0chain/issues/765 we need to integrate with[ this branch of gosdk](https://github.com/0chain/gosdk/tree/feature/transaction-nonce).

### PR DESCRIPTION
**Pull request by adityamatt**
- Caching of nonce for 5 minutes
- Sends nonce from getBalance to submitTransaction
- Closes #116 

To support https://github.com/0chain/0chain/issues/765

### CODE REVIEW

**Comment by adityamatt** [comment]
@shaktals  not sure who I should request to review the PR, can you tag the relevant person for review ?

**Comment by shaktals** [comment]
I can review it

File: index.js [hunk 1]
```diff
@@ -273,25 +275,7 @@ module.exports = {
         });
     },
 
-    getBalance: (client_id) => {
```
Comment (shaktals) [inline index.js:276]: Any specific reason to move the getBalance method?

File: index.js [hunk 2]
```diff
@@ -1178,23 +1155,53 @@ function createWalletKeys(mnemonic) {
...
+    }
+    const nonceToUse = cachedNonce+1
+    const hashdata = ts + ":" + nonceToUse + ":" + ae.id + ":" + toClientId + ":" + val + ":" + hashPayload;
     const hash = sha3.sha3_256(hashdata);
     const bytehash = utils.hexStringToByte(hash);
     const sec = new bls.SecretKey();
     sec.deserializeHexStr(ae.secretKey);
     const sig = sec.sign(bytehash);
 
+
```
Comment (shaktals) [inline index.js:1198]:
```suggestion
```

File: index.js [hunk 3]
```diff
@@ -1178,23 +1155,53 @@ function createWalletKeys(mnemonic) {
...
+        //initialize by 0 if there is no nonce from getBalance as well
+        try {
+            const { nonce } = await getBalance(ae.id)
+            cachedNonce = nonce ?? 0   
+        }
+        catch (err) {
+           cachedNonce = 0  
+        }
+    }
+    const nonceToUse = cachedNonce+1
```
Comment (shaktals) [inline index.js:1190]:
```suggestion
    }

    const nonceToUse = cachedNonce+1
```

File: index.js [hunk 4]
```diff
@@ -1178,23 +1155,53 @@ function createWalletKeys(mnemonic) {
...
+        });
+}
+
 async function submitTransaction(ae, toClientId, val, note, transaction_type) {
     const hashPayload = sha3.sha3_256(note);
     const ts = Math.floor(new Date().getTime() / 1000);
-
-    const hashdata = ts + ":" + ae.id + ":" + toClientId + ":" + val + ":" + hashPayload;
-
+    if (cachedNonce === undefined) {
```
Comment (shaktals) [inline index.js:1180]:
```suggestion

    if (cachedNonce === undefined) {
```

File: index.js [hunk 5]
```diff
@@ -42,6 +43,7 @@ const FaucetSmartContractAddress = '6dba10422e368813802877a85039d3985d96760ed844
 const StorageSmartContractAddress = '6dba10422e368813802877a85039d3985d96760ed844092319743fb3a76712d7';
 const MinerSmartContractAddress = '6dba10422e368813802877a85039d3985d96760ed844092319743fb3a76712d9';
 const InterestPoolSmartContractAddress = 'cf8d0df9bd8cc637a4ff4e792ffe3686da6220c45f0e1103baa609f3f1751ef4';
+const cachedNonceTime = 5*60*1000       //5 Minutes into milli seconds
```
Comment (shaktals) [inline index.js:46]: A more descriptive variable name would be something like:

```suggestion
const cachedNonceResetInterval = 5*60*1000       //5 Minutes into milli seconds
```

File: index.js [hunk 6]
```diff
@@ -1207,10 +1214,16 @@ async function submitTransaction(ae, toClientId, val, note, transaction_type) {
...
             })
             .catch((error) => {
                 reject(error);
             })
     });
 }
+
+setInterval(() => {
+    cachedNonce = undefined
+},cachedNonceTime)
```
Comment (shaktals) [inline index.js:1229]:
```suggestion
}, cachedNonceTime)

```

**Review by shaktals — CHANGES_REQUESTED** [review]
Looks good, before we merge and publish we need to:

- upgrade the lib version on package,json
- test this update with the related block-explorer updates, to confirm it all works together nicely

File: index.js (hunk 1 above)
Comment (adityamatt) [reply index.js:276]: I need to use getBalance to get the nonce to initialize, hence had to move it outside

**Comment by adityamatt** [comment]
> * test this update with the related block-explorer updates, to confirm it all works together nicely

We don't need any changes in block-explorer as explained here [Comment](https://github.com/0chain/block-explorer/issues/639#issuecomment-1107439165)

**Comment by adityamatt** [comment]
@cnlangzi  I think the zcn.wasm has some error. Following error is thrown
image

Steps to Reproduce :
- clone js-client-sdk locally and switch to this branch
- clone block-explorer and npm link the js-client-sdk local repo to be used by block-explorer
- Run and try uploading the file

Based on the error messages thrown. Do you think this is something thats coming from zcn.wasm

**Review by shaktals — APPROVED** [review]
lgtm

**Comment by shaktals** [comment]
@adityamatt remember that to run the `nonce` feature, you need to use the `hightps-2.devnet-0chain.net` network. 

I just did several file operations there, and everything seems to be working perfectly.

Now we just have to coordinate with @dabasov before we merge this and update all main networks.

My suggestion is to migrate one at a time.

**Comment by adityamatt** [comment]
aah. I was told to use alpha.devnet. 
Thanks for taking care of it.
@dabasov You can start updating the network I think.

**Comment by shaktals** [comment]
> @shaktals the PRs for nonce in backend is merged as per @dabasov . I think we can merge this now. But I think we'll also have to update the block-explorer package.json file to use the latest version of js-client-sdk or that happens automatically ?

Yeah, let's wait for @dabasov to confirm here that we are ready to go.
As for how NPM works, I've answered your question before, please research, very easy to find out.

**Pull request merged**